### PR TITLE
optimize AI

### PR DESCRIPTION
**Pull request by mjh0907**

### CODE REVIEW

**Comment by mjh0907** [comment]
Just FYI. Do not need to merge now.

**Comment by liuhb86** [comment]
Merged!

File: app/com/simplex9/splendor/actiongenerator/BuyCardGenerator.scala [hunk 1]
```diff
@@ -26,7 +26,11 @@ class BuyCardGenerator(state : State, playerIndex: Int, estimator: ValueEstimato
     if (card == null || card.isSecret) return None // an empty space or a secret reserve card
 
     // cannot afford
-    if (cardValue.lack >= player.golds) return None
+    if (cardValue.lack > player.golds) return None
+
+    val cardValueCostRateGood = (card.point == 0) ||
```
Comment (liuhb86) [inline app/com/simplex9/splendor/actiongenerator/BuyCardGenerator.scala:31]: is it intentional to skip check for point==0 card?
Or use something like (card.point + 1)/card.price.sum > ...?

File: app/com/simplex9/splendor/actiongenerator/ReserveCardGenerator.scala [hunk 2]
```diff
@@ -12,6 +12,17 @@ class ReserveCardGenerator(state: State, playerIndex: Int, estimators: Array[Val
   def generate(): List[Action] = {
     val player = state.players(playerIndex)
     if (player.reserve.length >= Param.MAX_RESERVE) return Nil
+    var gameStage = 0
+    for(i <- (0 to Param.GAME_STAGE_CARD_NUM.length-1).reverse){
+      if(Param.GAME_STAGE_CARD_NUM(i) <= player.cards.length){
```
Comment (liuhb86) [inline app/com/simplex9/splendor/actiongenerator/ReserveCardGenerator.scala:17]: player.cards.length is always 5?

File: app/com/simplex9/splendor/solver/Solver.scala [hunk 3]
```diff
@@ -47,7 +52,7 @@ class Solver(state: State, playerIndex: Int, startPlayer: Int) {
 
     if (isLeaf) {
       var score = StateEvaluator.evaluate(state, playerIndex)
-      if (isMinNode) score = -score
```
Comment (liuhb86) [inline app/com/simplex9/splendor/solver/Solver.scala:50]: ???

File: app/com/simplex9/splendor/actiongenerator/ReserveCardGenerator.scala (hunk 2 above)
Comment (mjh0907) [reply app/com/simplex9/splendor/actiongenerator/ReserveCardGenerator.scala:17]: oh...bug...:(

File: app/com/simplex9/splendor/solver/Solver.scala (hunk 3 above)
Comment (mjh0907) [reply app/com/simplex9/splendor/solver/Solver.scala:50]: why do we need this?

Comment (liuhb86) [reply app/com/simplex9/splendor/solver/Solver.scala:50]: Oh, yes, it's a bug!!!

File: app/com/simplex9/splendor/actiongenerator/BuyCardGenerator.scala [hunk 4]
```diff
@@ -26,7 +26,11 @@ class BuyCardGenerator(state : State, playerIndex: Int, estimator: ValueEstimato
     if (card == null || card.isSecret) return None // an empty space or a secret reserve card
 
     // cannot afford
-    if (cardValue.lack >= player.golds) return None
```
Comment (liuhb86) [inline app/com/simplex9/splendor/actiongenerator/BuyCardGenerator.scala:29]: OMG -_-!

**Pull request closed without merging**